### PR TITLE
Represent and run a minion model in Rust

### PR DESCRIPTION
**Pull request by niklasdewally**
I have a model running in Minion from a rust representation.

There are example problems for both the low level bindings and the Minion Rust AST versions inside raw_bindings.rs and run.rs respectively.

The work is very incomplete, things I still need to do (that I know about) are:

* How to deal with boolean / non int constants
* ~~Memory management / freeing C++ stuff~~
* I assume that variables used in the ast are defined in the symbol table, and if they aren't this could segfault.
* Return Result instead of panicking
*  Callbacks and Solutions
* Style and docs

### CODE REVIEW

**Comment by niklasdewally** [comment]
Feedback would be appreciated  @ozgurakgun @ChrisJefferson @lixitrixi 

(This can wait till Monday, I know I am working unsociable hours this week)

**Comment by niklasdewally** [comment]
No idea if the CI will work or not - as described in #27 minion runs single threaded, but cargo tests multithreaded

Tests run on host-server, haven't tested it on mac yet

File: solvers/minion/build.rs [hunk 1]
```diff
@@ -1,7 +1,8 @@
 // adapted from
-// - https://github.com/gokberkkocak/rust_glucose/blob/master/build.rs
+// e https://github.com/gokberkkocak/rust_glucose/blob/master/build.rs
```
Comment (ChrisJefferson) [inline solvers/minion/build.rs:2]: is this just a typo?

**Comment by ChrisJefferson** [comment]
I think this is quite good, and it would be nice to merge, because while lots of things still need doing, it gives us all a base for hacking on, where things run "end to end".

**Comment by ChrisJefferson** [comment]
To deal specifically with the crashing issue, I'd suggest changing the start of `runMinion` in libwrapper.cpp to look like this:
```
std::mutex global_minion_lock;
ReturnCodes runMinion(SearchOptions& options, SearchMethod& args, ProbSpec::CSPInstance& instance, bool(*callback)(void)) {
  std::lock_guard<std::mutex> guard(global_minion_lock);
```

This just makes a global mutex, and then locks it while in runMinion. This stops multiple calls to runMinion running at once (obviously). C++ is quite good at cleaning up (the mutex should be automatically unlocked when this function exits), so this should create problems even if minion exits with an error.

**Comment by ozgurakgun** [comment]
or should I? this is still marked as a draft?

**Comment by niklasdewally** [comment]
> or should I? this is still marked as a draft?

As long as you are happy with this still referencing  my Dev branch of minion, it can be merged

**Comment by niklasdewally** [comment]
Might be good to hack over later

File: solvers/minion/src/ast.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+/// variables, but is used to allow the use of a constant in the place of a variable in a
+/// constraint.
+pub enum Var {
+    NameRef(VarName),
+    ConstantAsVar(i32),
+}
+
+pub enum Constant {
+    Bool(bool),
+    Discrete(i32),
```
Comment (ozgurakgun) [inline solvers/minion/src/ast.rs:40]: maybe rename this to int

File: solvers/minion/src/ast.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+#[derive(Copy, Clone)]
+pub enum VarType {
+    Bounded(i32, i32),
```
Comment (ozgurakgun) [inline solvers/minion/src/ast.rs:45]: add discrete and sparsebound

**Review by ozgurakgun — COMMENTED** [review]
some minor comments

**Comment by niklasdewally** [comment]
@ozgurakgun I've renamed those things suggested.

**Pull request merged**